src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  actions.forEach((action) => {
+    const previousState =
+      states.length > 0 ? states[states.length - 1] : { ...state };
+    let newState = { ...previousState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+
+      default:
+        break;
+    }
+
+    states.push(newState);
```

Comment:
According to the best practices and the checklist, it is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after the current action is processed. You are already doing this correctly.